### ISSUE TITLE
Slow upload speed for http server

### ISSUE THREAD

**Issue by devalexqt** [denoland/deno#13608]
I tested upload speed for varius file sizes: 10M, 100M, 1000M in camparison of `node.js` upload speed in `Deno` is about  2x-3x slower and on 1G file deno is 10x slower and use almost all CPU resources. 
What I'm doing wrong or how to improve upload speed??

Send test data to server(localhost) via curl:
```
time curl -H "Content-Type:application/octet-stream" --data-binary @/dir/dev/foo1000M.bin  http://127.0.0.1:8080/upload/foo.bin
```
Test file was created by command:
```
truncate -s 10M foo10M.bin
```

```
import { serve } from "https://deno.land/std@0.125.0/http/server.ts";
import { config } from "./modules/config.js"
import * as path from "https://deno.land/std/path/mod.ts"
import {
    writableStreamFromWriter,
    readableStreamFromReader,
} from "https://deno.land/std/streams/mod.ts";

const PORT=8080

async function handler(req){
    console.log(`>>>new request: ${req.method}, ${req.url}, path: ${req.pathname} params: ${req.searchParams}`)
    console.log("HEADERS:",req.headers)

    if (req.body) {
        console.log("==>hasBody!");
        const file = await Deno.open(path.join(config.upload_dir, "native_deno.bin"), { create: true, write: true })
        const writableStream = writableStreamFromWriter(file)
        await req.body.pipeTo(writableStream)
    }
    console.log(">>upload complete!")
    return new Response("upload complete!");
}

serve(handler, { port: PORT })
console.log(">>>server listen...",PORT)
```

Results:
```
for 10M file:
   Deno:  60 - 90ms      Node:  35-40ms

for 100M file:
    Deno:  500-600ms   Node:  260ms

for 1000M file:
    Deno:  20s-5m          Node: 1000-1700ms
```

**Comment by lucacasonato** [comment]
Please include your `node` code too.

**Comment by devalexqt** [comment]
This is my simple test  `node` code:
```
const express = require('express')
const morgan = require('morgan')
const fs = require('fs')
const path = require('path')
const stream = require('stream/promises')

const app = express()
const port = 8081

app.use(morgan('tiny'))

app.get('/', (req, res) => {
    res.send('Hello World! ' + new Date())
})

app.post("/upload/:filename", async (req, res) => {

    //check file name
    if (!req?.params?.filename || !/^[0-9a-zA-Z\.\-\_]+$/.test(req?.params?.filename)) {
        return res.send("Invalid file name!")
    }

    const fs_stream = fs.createWriteStream(path.join("/dir/www/upload", req.params?.filename), { flags: "w" })

    //stream body to file
    await stream.pipeline(req, fs_stream)
    res.send("File successfully uploaded to node server!")
    fs_stream.close()
})

app.listen(port, () => {
    console.log(`==>listening on port ${port}`)
})
```

**Comment by devalexqt** [comment]
Then I test uploading 1G file multiple times I see performance degrades.

**Comment by kitsonk** [comment]
Duplicate of #10157

**Comment by lucacasonato** [comment]
@kitsonk Maybe - I have some stuff to do on Monday, but I'll try to investigate this on Tuesday.

**Comment by lucacasonato** [comment]
@devalexqt So I can reproduce Deno being slower than Node here, but not catastrophically slower like your benchmark suggests. I measured 1GB uploads to the Node server at 1.24 secs, and the Deno upload at 3.40 secs. I am using your exact script and commands, except that I am writing to /tmp instead of cwd.

Is there anything else I should know to be able to reproduce your results? What Deno version are you on?

**Comment by devalexqt** [comment]
> @devalexqt So I can reproduce Deno being slower than Node here, but not catastrophically slower like your benchmark suggests. I measured 1GB uploads to the Node server at 1.24 secs, and the Deno upload at 3.40 secs. I am using your exact script and commands, except that I am writing to /tmp instead of cwd.
> 
> Is there anything else I should know to be able to reproduce your results? What Deno version are you on?

Try to upload 1g file multiple times in the row, every time it's slower and slower...

2x-3x slower that node, looks catastrophic in my eyes, something wrong.

**Comment by lucacasonato** [comment]
I'll try that, thanks.

> 2x-3x slower that node, looks catastrophic in my eyes, something wrong in deno code under the hood.

I agree it isn't great, but 5s is much faster than 5 minutes (60x faster) 😅

> Also, I had 2 cpu and 2g ram server.

Ah, I think this may be related. I'll take this into account.

**Comment by devalexqt** [comment]
Deno also uses almost all CPU resource during upload  1G file.

**Comment by devalexqt** [comment]
Any news?

**Comment by ry** [comment]
We've recently added the ability to use Web Streams directly with file system files. 

https://deno.com/blog/v1.19#files-network-sockets-and-stdio-are-now-native-web-streams

@devalexqt I'd be interested to see your benchmark again:
```js
async function handler(req) {
    console.log(`>>>new request: ${req.method}, ${req.url}, path: ${req.pathname} params: ${req.searchParams}`)
    console.log("HEADERS:",req.headers)

    if (req.body) {
        console.log("==>hasBody!");
        const file = await Deno.create(path.join(config.upload_dir, "native_deno.bin"))
        await req.body.pipeTo(file.writable)
    }
    console.log(">>upload complete!")
    return new Response("upload complete!");
}
```

**Comment by devalexqt** [comment]
Thanks, will try today.

**Comment by devalexqt** [comment]
Hi!
Tested new version:
```
deno --version
deno 1.19.0 (release, x86_64-unknown-linux-gnu)
v8 9.9.115.7
typescript 4.5.2
```
I tested 1G file uploading and with new version total uploading time is reduced from 16.5 to 5.3 seconds (3x improvement and result more stable if I repeat test multiple times). But `node.js` has 2.3 seconds! (2x faster) and uses 2x less CPU power on my VPS server: 2 CPU & 2G RAM. 
I also want to try to run test on 4-8 core CPU.

**Comment by devalexqt** [comment]
Hi!
Running the tests on another VPS with 4 CPUs and 8 GB of RAM gives exactly the same result.

**Comment by devalexqt** [comment]
Tested again on new  `deno` and `std` versions:
```
deno 1.19.2
std@0.128.0
```
Deno still 2x slower vs `node` : 2.3sec vs 5.5sec

**Comment by devalexqt** [comment]
```
deno 1.21.0
std@0.136.0
```
Deno still 2x slower vs node : 2.3sec vs 5.5sec

**Comment by devalexqt** [comment]
Tested new `deno` release:
```
deno 1.21.1
std@0.137.0
```
Deno is slightly faster than previous version 4.5s, but still 2x slower than `node` 2.3s

**Comment by devalexqt** [comment]
Same result as before...
```
deno 1.21.3
std@0.139.0
```

**Comment by devalexqt** [comment]
Same result as  before....
```
deno 1.22.0
std@0.140.0
```

**Comment by devalexqt** [comment]
Same result as before....
```
deno 1.22.1
std@0.141.0
```

**Comment by devalexqt** [comment]
Same result as before....
```
deno 1.23.0
std@0.144.0
```

**Comment by devalexqt** [comment]
Same result as before....
```
deno 1.24.0
std@0.149.0
```

**Comment by bartlomieju** [comment]
We are actively working on improving performance of HTTP server.

**Comment by devalexqt** [comment]
I'm tested same server but with new `Deno.serve()`, result is the same: 2x slower vs `node`.
```
deno 1.25.0 (release, x86_64-unknown-linux-gnu)
v8 10.6.194.5
typescript 4.7.4
```

**Comment by devalexqt** [comment]
Same slow results.
```
deno 1.25.1 (release, x86_64-unknown-linux-gnu)
v8 10.6.194.5
typescript 4.7.4
```

**Comment by marcosc90** [comment]
ReadableStreams are quite slow as compared to the old Reader/Writer interface, because the buffer is not being reused. Allocating a new `Uint8Array` on every chunk is one of the reasons of the slow performance on big uploads.

https://github.com/denoland/deno/blob/d7b27ed63bf74c19b0bc961a52665960c199c53a/ext/http/01_http.js#L383-L390

By reusing the buffer (not using ReadableStream) uploading a 2gb file takes around 3 seconds, the same upload using a ReadableStream, it takes 4.8 seconds on my computer.

I love having spec compliant `Response` and `Request`, but when dealing with a lot of data is nice to have some lower level APIs. I think we should expose lower level APIs for users that really need that speed. 

The same bottleneck happens in `fetch` when downloading large files.

**Comment by marcosc90** [comment]
I'm working on this, I'll try to have a PR by tomorrow.

**Comment by lucacasonato** [comment]
Xref #16046

**Comment by devalexqt** [comment]
Same slow results.
```
deno 1.26.1 (release, x86_64-unknown-linux-gnu)
v8 10.7.193.3
typescript 4.8.3
```

**Comment by ripanioan** [comment]
On Windows, the gap between deno and node is getting closer.
\
Results for uploading an 1GB file:

`node` ~ 3.5 sec 
`deno` ~ 3.8 sec 

```
deno 1.27.0 (release, x86_64-pc-windows-msvc)
v8 10.8.168.4
typescript 4.8.3
```

**Comment by devalexqt** [comment]
Little bit faster but still  2x slower vs node.
```
deno 1.27.0 (release, x86_64-unknown-linux-gnu)
v8 10.8.168.4
typescript 4.8.3
```

**Comment by coderbuzz** [comment]
Hi, any update?

**Comment by devalexqt** [comment]
```
deno 1.29.4 (release, x86_64-unknown-linux-gnu)
std  0.173.0
```
Same slow results...

**Comment by coderbuzz** [comment]
How it compared with Bun?

**Comment by aapoalas** [comment]
What's the current code you're running by the way? I think `writableStreamFromReader` etc. are no longer needed as there's now a `.readable` and `.writable` property in the `Deno.open` result: https://deno.land/api@v1.29.4?s=Deno.FsFile#prop_readable

**Comment by aapoalas** [comment]
I tested this with the `.writable` property from `Deno.open`. I had to use my browser as the sending side because CURL would always set Content-Length to 10, and if I set it manually the upload would hang (presumably from Deno waiting for more data while CURL would not send any more).

Using the browser I sent `107374182` bytes, that is about 103M, in 880-920 milliseconds. So, based on that it seems like with the new APIs allotted that use the fast streams feature behind the scenes, performance is now on par with Node.

**Comment by aapoalas** [comment]
Profiling can be done by setting up the server from the topic statement (change to use `Deno.serve` or some other API if preferred) and start the server with `deno run --unstable -A --inspect server.ts`. Then connect via Chrome profiler, start a performance recording and start a curl session to upload 1000MB or 1GB file (use `-X POST -T 1GB_file.bin` if you run into curl memory issues).

I'll come back with a proper profiling code and setup later today.

**Comment by aapoalas** [comment]
Here's the way I profiled this:

```ts
// test.ts
import { dirname, fromFileUrl, join } from "https://deno.land/std/path/mod.ts";

const directory = dirname(fromFileUrl(import.meta.url));

Deno.serve(async (req) => {
  console.group("Request to", req.method, req.url);
  console.log(`Headers:`, req.headers);

  if (req.body) {
    console.log("Initiating upload");
    const start = performance.now();
    const file = await Deno.open(join(directory, "native_deno.bin"), {
      create: true,
      write: true,
    });
    await req.body.pipeTo(file.writable);
    console.log("Finished upload in", performance.now() - start, "ms");
  }
  console.groupEnd();
  return new Response();
});
```
And start the server with `deno run --inspect --unstable --allow-net=0.0.0.0 --allow-write=. test.ts`. Then start up a debugger with either Chrome / Chromium, or VS Code or some other way if that's your preference.

Create a 1 GB file with
```sh
truncate -s 1G foo1GB.bin
```
then start a Performance recording in debugger and send the file to the server with
```sh
time curl -H "Content-Type:application/octet-stream" -XPOST -T foo1G.bin  http://127.0.0.1:8000
```

### Side note about profiling performance:
Locally currently when I do not have a debugger connected, the 1GB upload finishes in just about a second.

When I connected a debugger but do not start up a performance recording, the upload takes anything from 3.5 to more than 4 seconds.

If I start a performance recording, the upload again goes down to about a second. This may indicate some issue with Deno regarding a connected-but-idle debugger client.

**Comment by rsgilbert** [comment]
Any update?

**Comment by devalexqt** [comment]
Upload 1G file:
```
deno 1.34.3 (release, x86_64-unknown-linux-gnu)
v8 11.5.150.2
typescript 5.0.4
std 0.192.0
```
`deno`:  2.7s
`node`:  1.8s

**Comment by lino-levan** [comment]
ref: #19737

**Comment by bartlomieju** [comment]
Revisiting this on the latest canary (`deno 2.8.2+ecc3053`) vs Node 24.8.0, uploading a 1 GB file via `curl -T` to a server that pipes the request body to disk.

**TL;DR:** The catastrophic part of this report (1 GB taking "20s–5m", CPU saturation, and getting progressively slower on each repeat) is no longer reproducible. The "~2x slower than Node" part was an apples-to-oranges comparison: the original Deno script uses **Web Streams** (`req.body.pipeTo(file.writable)`) while the Node script uses **native Node streams** (`stream.pipeline(req, fs.createWriteStream)`). When the API is held constant, the runtime gap disappears.

I decomposed it into a 2×2 matrix (same API on both runtimes), 1 GB upload, typical warm time:

| | API | Runtime | Time |
|---|---|---|---|
| A | node:http + native fs streams | Node | ~0.52 s |
| B | node:http + native fs streams | Deno | ~0.57 s |
| C | node:http + Web Streams (`Readable.toWeb`/`Writable.toWeb`) | Node | ~1.05 s |
| D | node:http + Web Streams | Deno | ~0.76 s |
| E | `Deno.serve` + `file.writable` (Web Streams) | Deno | ~0.80 s |

What this isolates:

1. **Same API → no runtime gap.** B vs A (both native streams): 0.57 s vs 0.52 s, within noise. Deno's socket read / file write / I/O plumbing is on par with Node.
2. **The cost is the Web Streams abstraction, not the runtime.** On Node alone, A→C goes 0.52 s → 1.05 s just by switching the same sockets/files to Web Streams. Node pays this penalty too. This matches @marcosc90's earlier observation: Web Streams allocate a fresh `Uint8Array` per chunk instead of reusing a buffer, plus promise-based backpressure overhead.
3. **On the actual apples-to-apples comparison (Web Streams on both), Deno is faster than Node:** D vs C is 0.76 s vs 1.05 s, and `Deno.serve` (E) lands right alongside it.

So the original 2x figure came from comparing Deno's spec-compliant Web Streams against Node's hand-optimized native streams. With the API held constant, Deno is on par with Node (native) and ahead of Node (Web Streams). The only remaining delta is the inherent per-chunk allocation cost of the Web Streams spec, which is tracked separately as a streams-perf optimization (BYOB / buffer reuse), not an HTTP-server or runtime issue.

Closing as the reported pathology is resolved.

(Benchmarked on Apple Silicon over loopback with an SSD; absolute numbers will differ from the original 2-core VPS, but the relative decomposition holds.)
